### PR TITLE
Feature/performance pipeline

### PR DESCRIPTION
**Pull request by cshcomcom**
## Description
- This features is required in performance-test-piepline

### CODE REVIEW

File: worker_node/job/send.js [hunk 1]
```diff
@@ -62,36 +62,49 @@ class Send extends Base {
...
+    }
+    const stakingBasePath = `/staking/${appName}/${this.config.ainAddress}/0/stake`;
+    const stakingBaseValue = await this.#ain.db.ref(stakingBasePath).getValue();
+    if (stakingBaseValue === null) {
+      const stakingTx = {
+        operation: {
+          type: 'SET_VALUE',
+          ref: `${stakingBasePath}/${Date.now()}/value`,
+          value: 1,
+          is_global: true,
```
Comment (platfowner) [inline worker_node/job/send.js:78]: question: is sharding enabled in the target nodes?
If not, this flag is skippable?

File: worker_node/job/send.js [hunk 2]
```diff
@@ -129,7 +142,7 @@ class Send extends Base {
   }
 
   async sendTxs() {
-    const delayTime = this.config.duration / this.config.numberOfTransactions * 1000;
+    const delayTime = (this.config.duration / this.config.numberOfTransactions * 1000) * 0.97;
```
Comment (platfowner) [inline worker_node/job/send.js:145]: Could you define 0.97 as a constant variable for readability?

File: worker_node/job/send.js [hunk 3]
```diff
@@ -62,36 +62,49 @@ class Send extends Base {
...
-    if (_.get(stakingTxResult, 'result.code') !== 0) {
-      throw Error(`Error while write staking tx (${JSON.stringify(stakingTxResult)})`);
+    const appName = (this.config.transactionOperation.ref).split('/')[2];
+    if (!appName) {
+      throw Error(`Can't find appName from transactionOperation.ref ` +
+          `(${this.config.transactionOperation.ref})`);
+    }
+    const stakingBasePath = `/staking/${appName}/${this.config.ainAddress}/0/stake`;
+    const stakingBaseValue = await this.#ain.db.ref(stakingBasePath).getValue();
+    if (stakingBaseValue === null) {
```
Comment (liayoo) [inline worker_node/job/send.js:72]: Fyi, if you want to check if there's some AIN staked for this app, you could also check if getValue at `/staking/${appName}/balance_total` > 0 !

File: worker_node/job/send.js (hunk 1 above)
Comment (cshcomcom) [reply worker_node/job/send.js:78]: Performance-test-pipeline disables sharding, but ain-benchmark supports shard node testcase.

File: worker_node/job/send.js (hunk 3 above)
Comment (cshcomcom) [reply worker_node/job/send.js]: Cool!!!! Done

File: worker_node/job/send.js (hunk 2 above)
Comment (cshcomcom) [reply worker_node/job/send.js]: Done!

**Comment by cshcomcom** [comment]
All comments are addressed! PTAL @platfowner @liayoo

File: worker_node/job/send.js [hunk 4]
```diff
@@ -129,7 +143,8 @@ class Send extends Base {
   }
 
   async sendTxs() {
-    const delayTime = this.config.duration / this.config.numberOfTransactions * 1000;
+    const delayAdjustmentRate = 0.97;
+    const delayTime = (this.config.duration / this.config.numberOfTransactions * 1000) * delayAdjustmentRate;
```
Comment (liayoo) [inline worker_node/job/send.js:147]: As discussed offline, let's try removing this `delayAdjustmentRate` and instead measure the more exact transmission times in the promises! Thanks for the discussion. If it's too much hassle, we can skip for now though.

File: worker_node/job/send.js (hunk 1 above)
Comment (platfowner) [reply worker_node/job/send.js:79]: I mean `stakingAppPath` does not seem to include any sharding path?

Comment (cshcomcom) [reply worker_node/job/send.js:79]: Ah.. Now I understand! It doesn't include any sharding path and can be removed!

Comment (cshcomcom) [reply worker_node/job/send.js]: Thanks done!

**Review by platfowner — APPROVED** [review]
LGTM except lia's comments.
Thanks!

File: worker_node/job/send.js (hunk 4 above)
Comment (cshcomcom) [reply worker_node/job/send.js]: Done! I added `targetTestEndTime` and use it in delay!

**Comment by cshcomcom** [comment]
PTAL 2! @liayoo

File: worker_node/job/send.js [hunk 5]
```diff
@@ -141,8 +153,10 @@ class Send extends Base {
     // const baseTimestamp = this.config.timestamp;
     const baseTx = this.makeBaseTransaction();
     const timestampSet = new Set();
+    const targetTestEndTime = Date.now() + (this.config.duration * 1000); // MS
 
     for (let i = 0; i < this.config.numberOfTransactions; i++) {
+      const delayTime = (targetTestEndTime - Date.now()) / (this.config.numberOfTransactions - i);
       await delay(delayTime);
```
Comment (liayoo) [inline worker_node/job/send.js:160]: Nice! Such an elegant solution! 👍 Could you just add a if (delayTime) > 0 condition check before calling delay()?

**Review by liayoo — APPROVED** [review]
LGTM with a comment. Thanks!

File: worker_node/job/send.js (hunk 5 above)
Comment (cshcomcom) [reply worker_node/job/send.js]: Thanks! Done!

**Comment by cshcomcom** [comment]
Thanks for the reviews! Now merging

**Pull request merged**